core/src/avm2/globals/flash/system/capabilities.rs

```diff
@@ -31,3 +31,43 @@ pub fn get_player_type<'gc>(
 
     Ok(AvmString::new_utf8(activation.context.gc_context, player_type).into())
 }
+
+/// Implements `flash.system.Capabilities.screenResolutionX`
+pub fn get_screen_resolution_x<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    _this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let viewport_dimensions = activation.context.renderer.viewport_dimensions();
+    Ok(viewport_dimensions.width.into())
+}
+
+/// Implements `flash.system.Capabilities.screenResolutionY`
+pub fn get_screen_resolution_y<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    _this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let viewport_dimensions = activation.context.renderer.viewport_dimensions();
+    Ok(viewport_dimensions.height.into())
+}
+
+/// Implements `flash.system.Capabilities.pixelAspectRatio`
+pub fn get_pixel_aspect_ratio<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    _this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    let viewport_dimensions = activation.context.renderer.viewport_dimensions();
+    Ok(viewport_dimensions.scale_factor.into())
```

Comment:
I am not certain that this is correct, but I also do not understand this part of the codebase.

https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/Capabilities.html#pixelAspectRatio

https://en.wikipedia.org/wiki/Pixel_aspect_ratio

Is `scale_factor` really the same as the "pixel aspect ratio"/`pixelAspectRatio`? This is the documentation I found for `scale_factor`:

https://github.com/ruffle-rs/ruffle/blob/master/render/src/backend.rs#L457-L459 :

```rust
    /// The scale factor of the containing viewport from standard-size pixels
    /// to device-scale pixels.
    pub scale_factor: f64,
```

I also found this page, which may or may not be relevant and helpful: https://help.adobe.com/en_US/as3/dev/WS5b3ccc516d4fbf351e63e3d118a9b90204-7cc6.html .

Reply:
since scale_factor is devicePixelRatio in web I assumed that this was pixelAspectRatio in flash

Reply:
I think `devicePixelRatio` comes from: https://developer.mozilla.org/en-US/docs/Web/API/Window/devicePixelRatio .

Also, I believe that `devicePixelRatio` describes some ratio between "physical" and "visual" pixels.

On the other hand, the "aspect ratio" in `pixelAspectRatio` describes (I believe) the relationship between width and height of some entity, in this case I believe the "aspect ratio" between the width and the height of a single pixel.

So, I would believe that these are two different abstractions.

Reply:
I see, `pixelAspectRatio` seems to be referring to the actual pixel ratio itself

Reply:
I would believe that a decent first implementation of `devicePixelRatio` would simply be `1`, since I do not know if Ruffle even supports any other ratio than `1`, and the capability in https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/Capabilities.html#pixelAspectRatio is immutable. But, it would be best if someone else other than me weighed in. Does anyone have any thoughts on this?

I would propose that you simply change it to that. See also my comment on the Ruffle forum.